File: proofs/fsharp_demo/lib/MathUtils.fs
// MathUtils.fs — real F# functions used as capture/validate targets.
//
// These are pure functions that take JSON-deserializable input and return
// JSON-serializable output. The Regrets harness passes input as a JSON
// string via env var, deserializes to JsonElement (boxed as obj), and
// calls the entry function. Entry must accept obj (the JsonElement) and
// return a JSON-serializable value.
//
// Three functions exported:
//   - add: takes obj (JsonElement representing a number), returns int
//   - classifyBmi: takes obj (JsonElement representing a number), returns string
//   - summarizeCart: takes obj (JsonElement representing an array of items),
//     returns a map { subtotal, tax, total, itemCount }

module MathUtils

open System
open System.Text.Json

/// Helper: extract a number from a JsonElement (handles int or float).
let private num (o: obj) : float =
    match o with
    | :? JsonElement as el ->
        match el.ValueKind with
        | JsonValueKind.Number -> el.GetDouble()
        | _ -> 0.0
    | :? int as i -> float i
    | :? float as f -> f
    | :? double as d -> d
    | _ -> 0.0

/// add: takes a JsonElement representing a number, returns the number +1.
/// (Simple increment function — good fingerprint target.)
let add (input: obj) : int =
    int (num input) + 1

/// classifyBmi: takes a JsonElement representing a BMI number, returns
/// a WHO category string.
let classifyBmi (input: obj) : string =
    let bmi = num input
    if bmi < 18.5 then "Underweight"
    elif bmi < 25.0 then "Normal"
    elif bmi < 30.0 then "Overweight"
    else "Obese"

/// summarizeCart: takes a JsonElement representing an array of cart items
/// (each { name, price, qty }), returns a map with subtotal, tax (11% PPN),
/// total, and itemCount.
let summarizeCart (input: obj) : obj =
    let items =
        match input with
        | :? JsonElement as el when el.ValueKind = JsonValueKind.Array ->
            el.EnumerateArray() |> Seq.toList
        | _ -> []
    let mutable subtotal = 0
    let mutable itemCount = 0
    for item in items do
        let price = item.GetProperty("price").GetInt32()
        let qty = item.GetProperty("qty").GetInt32()
        subtotal <- subtotal + (price * qty)
        itemCount <- itemCount + qty
    let tax = int (round (float subtotal * 0.11))
    let total = subtotal + tax
    // Return as a dict (will be JSON-serialized)
    dict [
        "subtotal", box subtotal
        "tax", box tax
        "total", box total
        "itemCount", box itemCount
    ] :> obj

/// formatRupiah: takes a JsonElement representing an integer amount,
/// returns a formatted Indonesian Rupiah string (e.g., 1500000 → "Rp 1.500.000").
let formatRupiah (input: obj) : string =
    let amount = int (num input)
    let absAmount = if amount < 0 then -amount else amount
    let s = string absAmount
    let buf = Text.StringBuilder()
    let mutable count = 0
    for i in (s.Length - 1) .. -1 .. 0 do
        if count > 0 && count % 3 = 0 then buf.Append('.') |> ignore
        buf.Append(s.[i]) |> ignore
        count <- count + 1
    let formatted =
        // Reverse
        let arr = buf.ToString().ToCharArray()
        Array.Reverse(arr)
        String(arr)
    if amount < 0 then "-Rp " + formatted
    else "Rp " + formatted

File: examples/dart/lib/math_utils.dart
// math_utils.dart — real Dart functions used as capture/validate targets.
//
// These are intentionally non-trivial so the fingerprint actually exercises
// code paths (not just constant returns). Two functions are exported:
//
//   - formatRupiah(int): String — format integer as Indonesian Rupiah.
//       e.g. 1500000 -> "Rp 1.500.000"
//
//   - classifyBmi(double): String — classify BMI per WHO categories.
//       e.g. 22.5 -> "Normal", 17.0 -> "Underweight"
//
//   - summarizeCart(List<Map>): Map — given a list of cart items
//       {name, price, qty}, compute subtotal, tax (11% PPN), total.
//       Returns: {subtotal: int, tax: int, total: int, itemCount: int}
//
// We use these because they touch: string formatting, branching on
// numeric ranges, and aggregate computation over a list of maps. A
// refactor that changes any of these will produce a different output
// and therefore a different fingerprint.

/// Format an integer as Indonesian Rupiah currency string.
/// Uses dot as thousand separator per Indonesian convention.
String formatRupiah(int amount) {
  if (amount < 0) return '-Rp ${_formatAbs(amount.abs())}';
  return 'Rp ${_formatAbs(amount)}';
}

String _formatAbs(int n) {
  final s = n.toString();
  final buf = StringBuffer();
  var count = 0;
  for (var i = s.length - 1; i >= 0; i--) {
    if (count > 0 && count % 3 == 0) buf.write('.');
    buf.write(s[i]);
    count++;
  }
  return String.fromCharCodes(buf.toString().codeUnits.reversed);
}

/// Classify BMI per WHO categories.
/// Returns one of: "Underweight", "Normal", "Overweight", "Obese".
String classifyBmi(double bmi) {
  if (bmi < 18.5) return 'Underweight';
  if (bmi < 25.0) return 'Normal';
  if (bmi < 30.0) return 'Overweight';
  return 'Obese';
}

/// Summarize a cart of items.
/// Each item: {"name": String, "price": int, "qty": int}.
/// Returns: {"subtotal": int, "tax": int, "total": int, "itemCount": int}
/// Tax is 11% PPN (Indonesian VAT), rounded to nearest integer.
///
/// Input is `List<dynamic>` (instead of `List<Map<...>>`) because the regret
/// harness passes JSON-decoded data, which always produces `List<dynamic>`
/// and `Map<String, dynamic>` regardless of the source-code type annotations.
/// We cast each element to `Map<String, dynamic>` inside the loop.
Map<String, Object> summarizeCart(List<dynamic> items) {
  var subtotal = 0;
  var itemCount = 0;
  for (final item in items) {
    final m = item as Map<String, dynamic>;
    final price = (m['price'] as num).toInt();
    final qty = (m['qty'] as num).toInt();
    subtotal += price * qty;
    itemCount += qty;
  }
  // 11% PPN, rounded to nearest integer
  final tax = (subtotal * 0.11).round();
  final total = subtotal + tax;
  return {
    'subtotal': subtotal,
    'tax': tax,
    'total': total,
    'itemCount': itemCount,
  };
}
